File: aoc-py/solutions/day7.py
"""
Day 7: Laboratories

https://adventofcode.com/2025/day/7
"""
__all__ = ('Day7',)

from typing import ClassVar

from ..solution import Solution

class Day7(Solution):
    NAME: ClassVar[str] = 'Laboratories'

    def part_one(self, inp: str) -> int:
        grid = [list(row) for row in inp.splitlines()]

        n_rows = len(grid)
        n_cols = len(grid[0])

        split = 0

        for i in range(1, n_rows):
            row = grid[i]

            for j in range(n_cols):
                if grid[i - 1][j] in ('S', '|'):
                    if row[j] == '.':
                        row[j] = '|'
                    elif row[j] == '^':
                        split += 1

                        if j > 0:
                            row[j - 1] = '|'
                        if j < n_cols:
                            row[j + 1] = '|'
        return split

    def part_two(self, inp: str) -> int:
        # using the same simulation algorithm as P1
        # instead we replace '.' and '|' with the (# of paths) from S to get to the current cell
        # i.e. empty cells '.' are never reached they will be 0
        # once reached a '|' will be a positive integer
        # use `-1` to represent splitters
        #
        # we will use the Pascal's triangle (binomial coefficient) unique paths counting algorithm
        # the (# of paths) to the current cell is the sum of the (# of paths) that reach this cell from the above row
        # i.e.
        # .|.|..| is actually .1.1..1
        # |^|^|.|             1^2^1.1
        grid = [
            [
                0 if cell == '.'
                else 1 if cell == 'S'
                else -1 for cell in row
            ]
            for row in inp.splitlines()
        ]

        n_rows = len(grid)
        n_cols = len(grid[0])

        for i in range(1, n_rows):
            row = grid[i]

            for j in range(n_cols):
                prev = grid[i - 1][j]

                # if previously there is no beam
                # then we will still have no beam...
                if prev > 0:
                    # if we are at a splitter
                    if row[j] == -1:
                        # we can propagate the # of paths onto the cells to the left and right (add them)
                        # we can guarantee that a splitter's (left/right) won't be other splitters
                        # if they are empty their value is `0` so we just add it on
                        # if they are not-empty and have been traversed to already (>0) i.e. from another splitter adjacent to it
                        # we just add onto the value of our beam (Pascal's triangle counting)
                        if j > 0:
                            row[j - 1] += prev
                        if j < n_cols:
                            row[j + 1] += prev
                    else:
                        # simply just propogate the # of paths from the same beam above
                        row[j] += prev

        # simply sum the (# of paths) to each final cell to get the total # of paths to the end
        return sum(grid[-1])

    def run(self, inp: str) -> None:
        print('Part 1:', p1 := self.part_one(inp))
        print('Part 2:', p2 := self.part_two(inp))

        assert p1 == 1630
        assert p2 == 47857642990160

File: src/bin/day7.rs
//! Day 7: Laboratories
//!
//! <https://adventofcode.com/2025/day/7>
use std::fmt::Display;
use aoc_2025::{Solution, get_grid};

pub struct Day7;

impl Solution for Day7 {
    const NAME: &'static str = "Laboratories";

    fn part_one<T: Display>(&self, inp: T) -> Self::OutputP1 {
        let mut grid = get_grid(inp);

        let n_rows = grid.len();
        let n_cols = grid[0].len();

        let mut split = 0;

        for i in 1..n_rows {
            for j in 0..n_cols {
                if grid[i - 1][j] == b'S' || grid[i - 1][j] == b'|' {
                    let row = &mut grid[i];

                    if row[j] == b'.' {
                        row[j] = b'|';
                    } else if row[j] == b'^' {
                        split += 1;

                        if j > 0 {
                            row[j - 1] = b'|';
                        }
                        if j < n_cols {
                            row[j + 1] = b'|';
                        }
                    }
                }
            }
        }
        split
    }

    fn part_two<T: Display>(&self, inp: T) -> Self::OutputP2 {
        let mut grid = inp
            .to_string()
            .lines()
            .map(|line| {
                line
                    .bytes()
                    .map(|cell| match cell {
                        b'.' => 0isize,
                        b'S' => 1,
                        _ => -1,
                    })
                    .collect::<Vec<_>>()
            })
            .collect::<Vec<_>>();

        let n_rows = grid.len();
        let n_cols = grid[0].len();

        for i in 1..n_rows {
            for j in 0..n_cols {
                let prev = grid[i - 1][j];
                let row = &mut grid[i];

                if prev > 0 {
                    if row[j] == -1 {
                        if j > 0 {
                            row[j - 1] += prev;
                        }
                        if j < n_cols {
                            row[j + 1] += prev;
                        }
                    } else {
                        row[j] += prev;
                    }
                }
            }
        }
        grid
            .last()
            .unwrap()
            .iter()
            .sum::<isize>()
            .cast_unsigned()
    }

    fn run(&self, inp: String) {
        let p1 = self.part_one(&inp);
        let p2 = self.part_two(&inp);

        println!("Part 1: {p1}");
        println!("Part 2: {p2}");

        assert_eq!(p1, 1630);
        assert_eq!(p2, 47_857_642_990_160);
    }
}

fn main() {
    aoc_2025::run_day(7, &Day7);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() { main(); }
}